cmd/gateway-s3.go

```diff
@@ -261,20 +261,32 @@ func fromMinioClientListBucketResult(bucket string, result minio.ListBucketResul
 // length indicates the total length of the object.
 func (l *s3Objects) GetObject(bucket string, key string, startOffset int64, length int64, writer io.Writer) error {
 	r := minio.NewGetReqHeaders()
-	if err := r.SetRange(startOffset, startOffset+length-1); err != nil {
-		return s3ToObjectError(traceError(err), bucket, key)
+
+	if length < 0 && length != -1 {
+		return s3ToObjectError(traceError(errInvalidArgument), bucket, key)
 	}
+	if length != -1 {
+		if err := r.SetRange(startOffset, startOffset+length-1); err != nil {
+			return s3ToObjectError(traceError(err), bucket, key)
+		}
+	}
+
 	object, _, err := l.Client.GetObject(bucket, key, r)
 	if err != nil {
 		return s3ToObjectError(traceError(err), bucket, key)
 	}
 
 	defer object.Close()
 
-	if _, err := io.CopyN(writer, object, length); err != nil {
-		return s3ToObjectError(traceError(err), bucket, key)
+	if length == -1 {
+		_, err = io.Copy(writer, object)
+	} else {
+		_, err = io.CopyN(writer, object, length)
```

Comment:
Actually we don't need `io.CopyN()` we already requested the right amount of file size from the backend we can just do `io.Copy()` instead - this will simplify the code further.